feat: improve how `ddev describe` or `ddev st` show exposed ports

hanoii opened the pull request:
This PR improves two things around ports:
 - provides ports in a list, rather and on a CSV, as projects with a lot of ports doesn't displays nicely.
 - gives a clear distinction between privately exposed ports (namely `InDocker` and the `Host` ports.

I've always been bothered by this. More often than not I want to clearly find what are the mappings of the local port to the remote ports. Also, onn some projects I exposed quite a bit of ports.

This is the second time I brought this up, couldn't find my previous issue though, but I hadn't worked on it back than.

On this project, for example, I exposed port 4000. As I have many ports I couldn't really tell  which host port was connected to container port 4000. I can find it with `docker ps` or some other inspection tools ,but it would be great if it was there and provided by ddev.

Somewhat related, this is the reason why I had to create https://github.com/hanoii/ddev-platformsh-lite/blob/main/commands/host/platform%3Atunnels on my not publicly exposed add-on.

Hope this doesn't takes as much as `ddev debug match-constraint` 😄 

Showing example of how it is displayed currently, vs the PR:

```
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: catalyze ~/Sites/confcats/catalyze https://catalyze.ddev.site                                                                                                                                                                                                                                                                     │
│ Docker platform: orbstack                                                                                                                                                                                                                                                                                                                  │
│ Router: traefik                                                                                                                                                                                                                                                                                                                            │
├──────────┬──────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE  │ STAT │ URL/PORT                                                                                                                                                                                                                                                                                            │ INFO               │
├──────────┼──────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────┤
│ web      │ OK   │ https://catalyze.ddev.site                                                                                                                                                                                                                                                                          │ drupal10 PHP8.2    │
│          │      │ InDocker: web:30006,30007,30010,30013,30018,4000,80,30001,30003,30011,30014,30015,30020,30004,30009,30016,443,8025,30019,30000,30002,30005,30008,30012,30017                                                                                                                                        │ nginx-fpm          │
│          │      │ Host: 127.0.0.1:32884,32884,32894,32894,32897,32897,32890,32890,32893,32893,32896,32896,32898,32898,32895,32895,32878,32878,32803,32879,32879,32881,32881,32882,32882,32887,32887,32888,32888,32891,32891,32892,32892,32899,32899,32880,32880,32883,32883,32885,32885,32886,32886,32889,32889,32804 │ docroot:'web'      │
│          │      │                                                                                                                                                                                                                                                                                                     │ Perf mode: mutagen │
│          │      │                                                                                                                                                                                                                                                                                                     │ NodeJS:18.18.2     │
├──────────┼──────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────┤
│ db       │ OK   │ InDocker: db:3306                                                                                                                                                                                                                                                                                   │ mariadb:10.11      │
│          │      │ Host: 127.0.0.1:32800                                                                                                                                                                                                                                                                               │ User/Pass: 'db/db' │
│          │      │                                                                                                                                                                                                                                                                                                     │ or 'root/root'     │
├──────────┼──────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────┤
│ Mailpit  │      │ Mailpit: https://catalyze.ddev.site:8026                                                                                                                                                                                                                                                            │                    │
│          │      │ Launch: ddev mailpit                                                                                                                                                                                                                                                                                │                    │
├──────────┼──────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────┤
│ All URLs │      │ https://catalyze.ddev.site, https://*.catalyze.ddev.site,                                                                                                                                                                                                                                           │                    │
│          │      │ https://127.0.0.1:32804, http://catalyze.ddev.site,                                                                                                                                                                                                                                                 │                    │
│          │      │ http://*.catalyze.ddev.site, http://127.0.0.1:32803                                                                                                                                                                                                                                                 │                    │
└──────────┴──────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────────────────┘
```

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: catalyze ~/Sites/confcats/catalyze https://catalyze.ddev.site                                   │
│ Docker platform: orbstack                                                                                │
│ Router: traefik                                                                                          │
├──────────┬──────┬───────────────────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE  │ STAT │ URL/PORT                                                          │ INFO               │
├──────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ web      │ OK   │ https://catalyze.ddev.site                                        │ drupal10 PHP8.2    │
│          │      │ InDocker: web:                                                    │ nginx-fpm          │
│          │      │   - 8025                                                          │ docroot:'web'      │
│          │      │ Host:                                                             │ Perf mode: mutagen │
│          │      │   - 127.0.0.1:32803->80                                           │ NodeJS:18.18.2     │
│          │      │   - 127.0.0.1:32804->443                                          │                    │
│          │      │   - 127.0.0.1:32878->4000                                         │                    │
│          │      │   - 127.0.0.1:32879->30000                                        │                    │
│          │      │   - 127.0.0.1:32880->30001                                        │                    │
│          │      │   - 127.0.0.1:32881->30002                                        │                    │
│          │      │   - 127.0.0.1:32882->30003                                        │                    │
│          │      │   - 127.0.0.1:32883->30004                                        │                    │
│          │      │   - 127.0.0.1:32884->30005                                        │                    │
│          │      │   - 127.0.0.1:32885->30006                                        │                    │
│          │      │   - 127.0.0.1:32886->30007                                        │                    │
│          │      │   - 127.0.0.1:32887->30008                                        │                    │
│          │      │   - 127.0.0.1:32888->30009                                        │                    │
│          │      │   - 127.0.0.1:32889->30010                                        │                    │
│          │      │   - 127.0.0.1:32890->30011                                        │                    │
│          │      │   - 127.0.0.1:32891->30012                                        │                    │
│          │      │   - 127.0.0.1:32892->30013                                        │                    │
│          │      │   - 127.0.0.1:32893->30014                                        │                    │
│          │      │   - 127.0.0.1:32894->30015                                        │                    │
│          │      │   - 127.0.0.1:32895->30016                                        │                    │
│          │      │   - 127.0.0.1:32896->30017                                        │                    │
│          │      │   - 127.0.0.1:32897->30018                                        │                    │
│          │      │   - 127.0.0.1:32898->30019                                        │                    │
│          │      │   - 127.0.0.1:32899->30020                                        │                    │
├──────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ db       │ OK   │ InDocker: db                                                      │ mariadb:10.11      │
│          │      │ Host:                                                             │ User/Pass: 'db/db' │
│          │      │   - 127.0.0.1:32800->3306                                         │ or 'root/root'     │
├──────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ Mailpit  │      │ Mailpit: https://catalyze.ddev.site:8026                          │                    │
│          │      │ Launch: ddev mailpit                                              │                    │
├──────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ All URLs │      │ https://catalyze.ddev.site, https://*.catalyze.ddev.site,         │                    │
│          │      │ https://127.0.0.1:32804, http://catalyze.ddev.site,               │                    │
│          │      │ http://*.catalyze.ddev.site, http://127.0.0.1:32803               │                    │
└──────────┴──────┴───────────────────────────────────────────────────────────────────┴────────────────────┘
```

hanoii commented:
edited description after my latest commit

stasadev reviewed (approved):
This is really amazing, thank you very much @hanoii !

It's much better than what we already have.

I tested it with installed `ddev/ddev-phpmyadmin` add-on. And used this `.ddev/docker-compose.ports.yaml`:

```yaml
services:
  web:
    expose:
      - 5492
      - 12394
      - 57497
    ports:
      - 5332:5222
      - 5555:5777
      - 12445
  db:
    expose:
      - 6594
      - 4352
    ports:
      - 12312:3999
      - 54355
  phpmyadmin:
    expose:
      - 8983
      - 12345
    ports:
      - 5000:5000
      - 5001:5605
      - 1234:14556
```

Usual test:

before
```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: d11 ~/code/ddev/d11 https://d11.ddev.site                                                   │
│ Docker platform: linux-docker                                                                        │
│ Router: traefik                                                                                      │
├────────────┬──────┬─────────────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                                    │ INFO               │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ web        │ OK   │ https://d11.ddev.site                                       │ php PHP8.3         │
│            │      │ InDocker: web:57497,5777,8025,443,12445,5222,5492,80,12394  │ nginx-fpm          │
│            │      │ Host: 127.0.0.1:5332,5332,5555,5555,33043,32795,32795,33044 │ docroot:''         │
│            │      │                                                             │ Perf mode: none    │
│            │      │                                                             │ NodeJS:20          │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ db         │ OK   │ InDocker: db:3306,3999,4352,54355,6594                      │ mysql:8.0          │
│            │      │ Host: 127.0.0.1:33045,12312,12312,32796,32796               │ User/Pass: 'db/db' │
│            │      │                                                             │ or 'root/root'     │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ phpmyadmin │ OK   │ https://d11.ddev.site:8037                                  │                    │
│            │      │ InDocker: phpmyadmin:80,80,8983,12345,14556,5000,5605       │                    │
│            │      │ Host: 127.0.0.1:1234,1234,5000,5000,5001,5001               │                    │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ Mailpit    │      │ Mailpit: https://d11.ddev.site:8026                         │                    │
│            │      │ Launch: ddev mailpit                                        │                    │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ All URLs   │      │ https://d11.ddev.site, https://127.0.0.1:33044,             │                    │
│            │      │ http://d11.ddev.site, http://127.0.0.1:33043                │                    │
└────────────┴──────┴─────────────────────────────────────────────────────────────┴────────────────────┘
```
after
```
┌─────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: d11 ~/code/ddev/d11 https://d11.ddev.site                                          │
│ Docker platform: linux-docker                                                               │
│ Router: traefik                                                                             │
├────────────┬──────┬────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                           │ INFO               │
├────────────┼──────┼────────────────────────────────────────────────────┼────────────────────┤
│ web        │ OK   │ https://d11.ddev.site                              │ php PHP8.3         │
│            │      │ InDocker:                                          │ nginx-fpm          │
│            │      │   - web:5492                                       │ docroot:''         │
│            │      │   - web:8025                                       │ Perf mode: none    │
│            │      │   - web:12394                                      │ NodeJS:20          │
│            │      │   - web:57497                                      │                    │
│            │      │ Host:                                              │                    │
│            │      │   - 127.0.0.1:33043->80                            │                    │
│            │      │   - 127.0.0.1:33044->443                           │                    │
│            │      │   - 127.0.0.1:5332->5222                           │                    │
│            │      │   - 127.0.0.1:5555->5777                           │                    │
│            │      │   - 127.0.0.1:32795->12445                         │                    │
├────────────┼──────┼────────────────────────────────────────────────────┼────────────────────┤
│ db         │ OK   │ InDocker:                                          │ mysql:8.0          │
│            │      │   - db:4352                                        │ User/Pass: 'db/db' │
│            │      │   - db:6594                                        │ or 'root/root'     │
│            │      │ Host:                                              │                    │
│            │      │   - 127.0.0.1:33045->3306                          │                    │
│            │      │   - 127.0.0.1:12312->3999                          │                    │
│            │      │   - 127.0.0.1:32796->54355                         │                    │
├────────────┼──────┼────────────────────────────────────────────────────┼────────────────────┤
│ phpmyadmin │ OK   │ https://d11.ddev.site:8037                         │                    │
│            │      │ InDocker:                                          │                    │
│            │      │   - phpmyadmin:80                                  │                    │
│            │      │   - phpmyadmin:8983                                │                    │
│            │      │   - phpmyadmin:12345                               │                    │
│            │      │ Host:                                              │                    │
│            │      │   - 127.0.0.1:5000->5000                           │                    │
│            │      │   - 127.0.0.1:5001->5605                           │                    │
│            │      │   - 127.0.0.1:1234->14556                          │                    │
├────────────┼──────┼────────────────────────────────────────────────────┼────────────────────┤
│ Mailpit    │      │ Mailpit: https://d11.ddev.site:8026                │                    │
│            │      │ Launch: ddev mailpit                               │                    │
├────────────┼──────┼────────────────────────────────────────────────────┼────────────────────┤
│ All URLs   │      │ https://d11.ddev.site, https://127.0.0.1:33044,    │                    │
│            │      │ http://d11.ddev.site, http://127.0.0.1:33043       │                    │
└────────────┴──────┴────────────────────────────────────────────────────┴────────────────────┘
```

Another test with `ddev config global --omit-containers="ddev-router"`

before
```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: d11 ~/code/ddev/d11 http://127.0.0.1:33021                                                  │
│ Docker platform: linux-docker                                                                        │
│ Router: disabled                                                                                     │
├────────────┬──────┬─────────────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                                    │ INFO               │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ web        │ OK   │ http://127.0.0.1:33021                                      │ php PHP8.3         │
│            │      │ InDocker: web:12394,12445,443,57497,5222,5492,5777,80,8025  │ nginx-fpm          │
│            │      │ Host: 127.0.0.1:33023,5332,5332,32782,32782,33021,5555,5555 │ docroot:''         │
│            │      │                                                             │ Perf mode: none    │
│            │      │                                                             │ NodeJS:20          │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ db         │ OK   │ InDocker: db:4352,54355,6594,3306,3999                      │ mysql:8.0          │
│            │      │ Host: 127.0.0.1:33022,12312,12312,32780,32780               │ User/Pass: 'db/db' │
│            │      │                                                             │ or 'root/root'     │
├────────────┼──────┼─────────────────────────────────────────────────────────────┼────────────────────┤
│ phpmyadmin │ OK   │ InDocker: phpmyadmin:12345,14556,5000,5605,80,80,8983       │                    │
│            │      │ Host: 127.0.0.1:5000,5000,5001,5001,8036,8036,1234,1234     │                    │
└────────────┴──────┴─────────────────────────────────────────────────────────────┴────────────────────┘
```
after
```
┌─────────────────────────────────────────────────────────────────────┐
│ Project: d11 ~/code/ddev/d11 http://127.0.0.1:33021                 │
│ Docker platform: linux-docker                                       │
│ Router: disabled                                                    │
├────────────┬──────┬────────────────────────────┬────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                   │ INFO               │
├────────────┼──────┼────────────────────────────┼────────────────────┤
│ web        │ OK   │ http://127.0.0.1:33021     │ php PHP8.3         │
│            │      │ InDocker:                  │ nginx-fpm          │
│            │      │   - web:5492               │ docroot:''         │
│            │      │   - web:8025               │ Perf mode: none    │
│            │      │   - web:12394              │ NodeJS:20          │
│            │      │   - web:57497              │                    │
│            │      │ Host:                      │                    │
│            │      │   - 127.0.0.1:33021->80    │                    │
│            │      │   - 127.0.0.1:33023->443   │                    │
│            │      │   - 127.0.0.1:5332->5222   │                    │
│            │      │   - 127.0.0.1:5555->5777   │                    │
│            │      │   - 127.0.0.1:32782->12445 │                    │
├────────────┼──────┼────────────────────────────┼────────────────────┤
│ db         │ OK   │ InDocker:                  │ mysql:8.0          │
│            │      │   - db:4352                │ User/Pass: 'db/db' │
│            │      │   - db:6594                │ or 'root/root'     │
│            │      │ Host:                      │                    │
│            │      │   - 127.0.0.1:33022->3306  │                    │
│            │      │   - 127.0.0.1:12312->3999  │                    │
│            │      │   - 127.0.0.1:32780->54355 │                    │
├────────────┼──────┼────────────────────────────┼────────────────────┤
│ phpmyadmin │ OK   │ InDocker:                  │                    │
│            │      │   - phpmyadmin:8983        │                    │
│            │      │   - phpmyadmin:12345       │                    │
│            │      │ Host:                      │                    │
│            │      │   - 127.0.0.1:8036->80     │                    │
│            │      │   - 127.0.0.1:5000->5000   │                    │
│            │      │   - 127.0.0.1:5001->5605   │                    │
│            │      │   - 127.0.0.1:1234->14556  │                    │
└────────────┴──────┴────────────────────────────┴────────────────────┘
```

Gitpod:

before
```
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: d10simple /workspace/code https://8080-ddev-ddevgitpodlaunch-9wh2a78be3v.ws-eu116.gitpod.io       │
│ Docker platform: linux-docker                                                                              │
│ Router: disabled                                                                                           │
├────────────┬──────┬───────────────────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                                          │ INFO               │
├────────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ web        │ OK   │ https://8080-ddev-ddevgitpodlaunch-9wh2a78be3v.ws-eu116.gitpod.io │ drupal PHP8.3      │
│            │      │ InDocker: web:80,12445,443,5222,5492,12394,57497,5777,8025        │ nginx-fpm          │
│            │      │ Host: 127.0.0.1:8080,8443,5332,5332,5555,5555,32770,32770,8027    │ docroot:'web'      │
│            │      │                                                                   │ Perf mode: none    │
│            │      │                                                                   │ NodeJS:20          │
├────────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ db         │ OK   │ InDocker: db:3306,3999,4352,54355,6594                            │ mariadb:10.11      │
│            │      │ Host: 127.0.0.1:3306,12312,12312,32769,32769                      │ User/Pass: 'db/db' │
│            │      │                                                                   │ or 'root/root'     │
├────────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ phpmyadmin │ OK   │ InDocker: phpmyadmin:5000,5605,80,80,8983,12345,14556             │                    │
│            │      │ Host: 127.0.0.1:1234,1234,5000,5000,5001,5001,8036,8036           │                    │
├────────────┼──────┼───────────────────────────────────────────────────────────────────┼────────────────────┤
│ Network    │      │ bind-all-interfaces ENABLED                                       │                    │
└────────────┴──────┴───────────────────────────────────────────────────────────────────┴────────────────────┘
```
after
```
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: d10simple /workspace/d10simple https://8080-hanoii-ddev-s6sreye647e.ws-eu116.gitpod.io  │
│ Docker platform: linux-docker                                                                    │
│ Router: disabled                                                                                 │
├────────────┬──────┬─────────────────────────────────────────────────────────┬────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                                │ INFO               │
├────────────┼──────┼─────────────────────────────────────────────────────────┼────────────────────┤
│ web        │ OK   │ https://8080-hanoii-ddev-s6sreye647e.ws-eu116.gitpod.io │ drupal PHP8.3      │
│            │      │ InDocker:                                               │ nginx-fpm          │
│            │      │   - web:5492                                            │ docroot:'web'      │
│            │      │   - web:12394                                           │ Perf mode: none    │
│            │      │   - web:57497                                           │ NodeJS:20          │
│            │      │ Host:                                                   │                    │
│            │      │   - 127.0.0.1:8080->80                                  │                    │
│            │      │   - 127.0.0.1:8443->443                                 │                    │
│            │      │   - 127.0.0.1:5332->5222                                │                    │
│            │      │   - 127.0.0.1:5555->5777                                │                    │
│            │      │   - 127.0.0.1:8027->8025                                │                    │
│            │      │   - 127.0.0.1:32771->12445                              │                    │
├────────────┼──────┼─────────────────────────────────────────────────────────┼────────────────────┤
│ db         │ OK   │ InDocker:                                               │ mariadb:10.11      │
│            │      │   - db:4352                                             │ User/Pass: 'db/db' │
│            │      │   - db:6594                                             │ or 'root/root'     │
│            │      │ Host:                                                   │                    │
│            │      │   - 127.0.0.1:3306->3306                                │                    │
│            │      │   - 127.0.0.1:12312->3999                               │                    │
│            │      │   - 127.0.0.1:32770->54355                              │                    │
├────────────┼──────┼─────────────────────────────────────────────────────────┼────────────────────┤
│ phpmyadmin │ OK   │ InDocker:                                               │                    │
│            │      │   - phpmyadmin:8983                                     │                    │
│            │      │   - phpmyadmin:12345                                    │                    │
│            │      │ Host:                                                   │                    │
│            │      │   - 127.0.0.1:8036->80                                  │                    │
│            │      │   - 127.0.0.1:5000->5000                                │                    │
│            │      │   - 127.0.0.1:5001->5605                                │                    │
│            │      │   - 127.0.0.1:1234->14556                               │                    │
├────────────┼──────┼─────────────────────────────────────────────────────────┼────────────────────┤
│ Network    │      │ bind-all-interfaces ENABLED                             │                    │
└────────────┴──────┴─────────────────────────────────────────────────────────┴────────────────────┘
```

rfay reviewed (changes requested):
I love this, it's an incredible improvement!

- [ ] Could you please check test coverage?
- [ ] Could you take a look at the "All URLs" and see if it might be possible to do something similar there? It's the exact same class of problem. Not trying to expand scope, but it does look like exactly the same problem. (The label is also ambigous, it's really just showing the primary http and https ports for the webserver.)
- [ ] Don't forget about the json version, `ddev describe -j`. This breaks `host_ports` and `exposed_ports`:
```
    "web": {
      "exposed_ports": "  - web:5492\n  - web:8025\n  - web:12394\n  - web:57497",
      "full_name": "ddev-d11-web",
      "host_http_url": "http://127.0.0.1:32872",
      "host_https_url": "https://127.0.0.1:32871",
      "host_ports": "  - 127.0.0.1:32872->80\n  - 127.0.0.1:32871->443\n  - 127.0.0.1:5332->5222\n  - 127.0.0.1:5555->5777\n  - 127.0.0.1:32774->12445",
      "http_url": "http://d11.ddev.site",
      "https_url": "https://d11.ddev.site",
      "image": "ddev/ddev-webserver:20241031_improve_php_image_build",
      "short_name": "web",
      "status": "running",
      "virtual_host": "d11.ddev.site"
    }
```

I think the previous version was more parse-able:
```
    "web": {
      "exposed_ports": "12445,57497,80,12394,443,5222,5492,5777,8025",
      "full_name": "ddev-d11-web",
      "host_http_url": "http://127.0.0.1:32872",
      "host_https_url": "https://127.0.0.1:32871",
      "host_ports": "32774,32774,5332,5332,32872,32871,5555,5555",
      "http_url": "http://d11.ddev.site",
      "https_url": "https://d11.ddev.site",
      "image": "ddev/ddev-webserver:20241031_improve_php_image_build",
      "short_name": "web",
      "status": "running",
      "virtual_host": "d11.ddev.site"
    }
  },
  ```

hanoii commented:
> Could you please check test coverage?

What do you mean, adding test? All tests passed and this is just presentation logic.

>  Could you take a look at the "All URLs" and see if it might be possible to do something similar there? It's the exact same class of problem. Not trying to expand scope, but it does look like exactly the same problem. (The label is also ambigous, it's really just showing the primary http and https ports for the webserver.)

I don't mind, but what do you have in mind? It's not the same problem because as you said what it's showing is at least clear enough. If I were to suggest something it could be:
- Remove it altogether as the main urls are shown on each section anyway.
- Change it to something Summary, or even leave All URLs but show a list with only the https versions of the sites with the domains. I wonder how many use it without https and with 127.0.0.1. 

I was checking a bit at the logic and saw the alternatives to with and without router and with and without http, but I would just show either based on that same logic rather than showing all. Further more I think (haven't looked properly) that the url that is shown on the section will change depending on those factors already, so again, I wonder if All URLs is necessary.

rfay commented:
> What do you mean, adding test? All tests passed and this is just presentation logic.

See if TestCmdDescribe can be slightly improved to check this usage. Or consider it. I'm not insisting. "this is just presentation logic" -- well, all of `ddev describe` is presentation :)

> I don't mind, but what do you have in mind? It's not the same problem because as you said what it's showing is at least clear enough. If I were to suggest something it could be:

> Remove it altogether as the main urls are shown on each section anyway.
> Change it to something Summary, or even leave All URLs but show a list with only the https versions of the sites with the domains. I wonder how many use it without https and with 127.0.0.1.

The router logic is quite different, this info can't be removed, it is *not* covered in each section. The "All URLs" has both by-URL and by-IP URLs, and they're unrelated to the ones in the web section, because they show what the router is doing, and that's the normal situation.

rfay commented:
The other great and obscure thing here is it makes that second column possibly shorter, resulting in less wrapping on smaller terminals.

hanoii commented:
- Added test coverage

>  Could you take a look at the "All URLs" and see if it might be possible to do something similar there? It's the exact same class of problem. Not trying to expand scope, but it does look like exactly the same problem. (The label is also ambigous, it's really just showing the primary http and https ports for the webserver.)

I changed All URLs for Project URLs, found it less missleading.. but we can not do that. Regardless I would do any further work on that, if any, on a separate issue.

@rfay question on testings.. The tests pass locally if I run `make testcmd TESTARGS="-run TestCmdDescribe"`

However, on Go Land it fails and I noticed the output is the previous version, like it's not using the built ddev but rather my local one?

rfay commented:
The tests in the cmd directory most often use the ddev in the path, yours is probably not this one. Docs have this case, 
https://ddev.readthedocs.io/en/stable/developers/building-contributing/#testing

> To run a test (in the cmd package) against a individually-compiled DDEV binary, set the DDEV_BINARY_FULLPATH environment variable, for example DDEV_BINARY_FULLPATH=$PWD/.gotmp/bin/linux_amd64/ddev make testcmd.

Thanks for your work on this! 

My normal working situation is I have a symlink in /usr/local/bin/ddev that normally points to my locally built version, but I use a script to point it to the homebrew version. So I'm almost always using the locally built version.

rfay reviewed (changes requested):
Just took a quick look, and I like it, but changing the entire format of the `ddev describe -j` `host_ports` is a breaking change I would think. I know it wouldn't affect many people, but we try not to do that. I don't think it's parseable either.

```
 "host_ports": "33025->80, 33024->443, 5332->5222, 5555->5777, 32780->12445",
```

hanoii commented:
The trailing comma I can fix. The other comment not sure what you want me
to do with it

Reply might be shorter than usual / Respuesta más resumida que de costumbre
Enviado desde el celular / Sent from Mobile


On Sat, 9 Nov 2024 at 17:53 Randy Fay ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Just took a quick look, and I like it, but changing the entire format of
> the ddev describe -j host_ports is a breaking change I would think. I
> know it wouldn't affect many people, but we try not to do that. I don't
> think it's parseable either.
>
>  "host_ports": "33025->80, 33024->443, 5332->5222, 5555->5777, 32780->12445",
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ddev/ddev/pull/6699#pullrequestreview-2425583180>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFFP53RKI4KHNTFLSSBD5LZ7ZY53AVCNFSM6AAAAABRLOJVS2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIMRVGU4DGMJYGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hanoii commented:
Ah, there's no trailing comma, you just copy over a portion of it, right? What did you mean with it's not parseable?

hanoii commented:
Tests pass!

rfay commented:
Sorry, I wasn't clear. This completely changes `host_ports` from a list of ports to a list of mappings. That's the problem, not trailing commas.

hanoii commented:
I understand..  well, your call here. I can go back to the previous version only for json output, or I can add a new element, host_ports_mapping and print that instead of that on the non json part.

hanoii commented:
```
        "host_ports": "32803,32804,32878,32879,32880,32881,32882,32883,32884,32885,32886,32887,32888,32889,32890,32891,32892,32893,32894,32895,32896,32897,32898,32899",
        "host_ports_mapping": "32803->80,32804->443,32878->4000,32879->30000,32880->30001,32881->30002,32882->30003,32883->30004,32884->30005,32885->30006,32886->30007,32887->30008,32888->30009,32889->30010,32890->30011,32891->30012,32892->30013,32893->30014,32894->30015,32895->30016,32896->30017,32897->30018,32898->30019,32899->30020",
```

rfay commented:
host_ports_mapping is a great idea, but it should probably be a json map instead of plain text "->"

hanoii commented:
I started working on this, is this oK:

```json
        "host_ports_mapping": [
          {
            "exposed_port": "80",
            "host_port": "32803"
          },
          {
            "exposed_port": "443",
            "host_port": "32804"
          },
          {
            "exposed_port": "4000",
            "host_port": "32878"
          },
          {
            "exposed_port": "30000",
            "host_port": "32879"
          },
```

And also, something I should mention that maybe you missed and just to make sure you are OK with it is that I also changed the content of exposed_ports.. previously it included the ones that are also exported to the host, now I am only shoing those that are not "only" exposed in the container. A mapped port to the port is always exposed in the contianer as well obviously. That was my reasoning. So one option is that you are ok with it and we leave it like that, or I can leave exposed ports for json the same way as before and do some extra logic on the table to remove them.

rfay commented:
I definitely like that for host_ports_mapping. Will look closely when I test at exposed_ports.

hanoii commented:
Worked on this

hanoii commented:
well, not sure why the static validation is failing. seems something completely out of this PR.

rfay commented:
I think you don't have the latest golangci-lint, you'll probably get the same results then. Newest is 1.62.0. I had 1.61.0. Annoyingly they keep "improving" its behavior over time :)

stasadev commented:
Rebased to fix the `golangci-lint` check.

rfay commented:
Something's wrong here. It shows only port web:8025 for inDocker, should be 80 (primary), 443, 8025.

```
$ ddev describe
┌───────────────────────────────────────────────────────────────────────────────────────┐
│ Project: d11 ~/workspace/d11 https://d11.ddev.site                                    │
│ Docker platform: orbstack                                                             │
│ Router: traefik                                                                       │
├──────────────┬──────┬────────────────────────────────────────────┬────────────────────┤
│ SERVICE      │ STAT │ URL/PORT                                   │ INFO               │
├──────────────┼──────┼────────────────────────────────────────────┼────────────────────┤
│ web          │ OK   │ https://d11.ddev.site                      │ drupal PHP8.3      │
│              │      │ InDocker:                                  │ apache-fpm         │
│              │      │   - web:8025                               │ docroot:'web'      │
│              │      │ Host:                                      │ Perf mode: mutagen │
│              │      │   - 127.0.0.1:32768->80                    │ NodeJS:20          │
│              │      │   - 127.0.0.1:32769->443                   │                    │
├──────────────┼──────┼────────────────────────────────────────────┼────────────────────┤
│ db           │ OK   │ InDocker: db                               │ mariadb:10.11      │
│              │      │ Host:                                      │ User/Pass: 'db/db' │
│              │      │   - 127.0.0.1:32770->3306                  │ or 'root/root'     │
├──────────────┼──────┼────────────────────────────────────────────┼────────────────────┤
│ Mailpit      │      │ Mailpit: https://d11.ddev.site:8026        │                    │
│              │      │ Launch: ddev mailpit                       │                    │
├──────────────┼──────┼────────────────────────────────────────────┼────────────────────┤
│ Project URLs │      │ https://d11.ddev.site,                     │                    │
│              │      │ https://127.0.0.1:32769,                   │                    │
│              │      │ http://d11.ddev.site,                      │                    │
│              │      │ http://127.0.0.1:32768                     │                    │
└──────────────┴──────┴────────────────────────────────────────────┴────────────────────┘
```

hanoii commented:
It's not "wrong" as in it's how I intended it.

This is how `docker ps` shows the ports:

```
8025/tcp, 127.0.0.1:33016->80/tcp, 127.0.0.1:33017->443/tcp
```

I basically did the same thing. On the InDocker part, you only see the ports that are "exclusively" exposed in docker (`expose` in docker compose lingo), and on the host you see all ports mapping from host to container (`ports` in docker compose lingo). There you have 80 and 443. For me, showing 80 and 443 on the InDocker part was completely redundant.

hanoii commented:
@rfay you'll obviously have the last call here, but I would like to insist in something like what I am suggesting because I `ddev st` a lot and adding back those ports to in decker would get me back to a broken table when there are a lot of ports or a very long table if I do it line by line and I duplicate things

So what about this, for the basic example (only a bare ddev and only showing the URL/PORT column of the web service):

Instead of:

```
https://a.ddev.site       
InDocker: web:443,80,8025 
Host: 127.0.0.1:33042,33041
```

what about:

```
https://a.ddev.site       
InDocker -> Host:
  - web:80 -> 127.0.0.1:33041
  - web:443 -> 127.0.0.1:33042
  - web:8025
 ```

Thus, combining the two, giving priority to what's on the container, but making it very clear which host port is mapped to it.

🤞🏻 you are OK with this.

hanoii commented:
Or:

```
InDocker <- Host:
  - web:80 <- 127.0.0.1:33041
  - web:443 <- 127.0.0.1:33042
  - web:8025
```

rfay commented:
Oh, those are improving things significantly! I like

```
https://a.ddev.site       
InDocker -> Host:
  - web:80 -> 127.0.0.1:33041
  - web:443 -> 127.0.0.1:33042
  - web:8025
```

quite a lot. And now it's not just providing information, it's providing context. I can go with this. Try not to break the json version.

Now if we could explain to people what's going on with the router too. Not sure they'll ever understand. (out of scope)

hanoii commented:
You prefer `->` to `<-`? The latter is more semantically correct :)

json will remain the same as before (I need to fix this)

rfay commented:
> You prefer -> to <-? The latter is more semantically correct :)

Well, that depends on which way you're looking at the traffic! 

`- web:80 -> 127.0.0.1:33041` seems to imply to me that port 80 is connected to 33041. Also, people are annoyed by backward arrows :)

hanoii commented:
I am curious on how many people use `ddev st -j`, do you have any stats on that?

rfay commented:
> I am curious on how many people use ddev st -j, do you have any stats on that?

No stats, I imagine it's few. Mostly probably wrapped in scripts that determine things about a project. There have been a few of those tossed around over the years.

stasadev reviewed (approved):
Looks great, even better!

And if it matters, I noticed that `host_ports` are not sorted:

Tested with config from https://github.com/ddev/ddev/pull/6699#pullrequestreview-2421667591

```
$ ddev st -j | jq .raw.services.phpmyadmin | grep 'ports"'
  "exposed_ports": "80,5000,5605,8983,12345,14556",
  "host_ports": "5000,5001,1234",

$ ddev st -j | jq .raw.services.web | grep 'ports"'
  "exposed_ports": "80,443,3000,5222,5492,5777,8025,12394,12445,57497",
  "host_ports": "32866,32867,5332,5555,32770",

$ ddev st -j | jq .raw.services.db | grep 'ports"'
  "exposed_ports": "3306,3999,4352,6594,54355",
  "host_ports": "32863,12312,32768",
```

hanoii commented:
> And if it matters, I noticed that host_ports are not sorted:

it mattered.. fixed that and added testing for being sort.

hanoii commented:
rebased

rfay commented:
@stasadev mentioned in our meeting today how absolutely thrilled he is with this ❤️

rfay reviewed (approved):
It's great! Thanks for this marvelous contribution!
